src/main/java/org/cryptomator/cryptofs/fh/ChunkCache.java

```diff
@@ -20,54 +25,166 @@ public class ChunkCache {
 	private final ChunkSaver chunkSaver;
 	private final CryptoFileSystemStats stats;
 	private final BufferPool bufferPool;
-	private final Cache<Long, Chunk> chunks;
+	private final ExceptionsDuringWrite exceptionsDuringWrite;
+	private final ConcurrentMap<Long, Chunk> staleChunks;
+	private final ConcurrentMap<Long, Chunk> activeChunks;
+
+	/**
+	 * This lock ensures no chunks are passed between stale and active state while flushing,
+	 * as flushing requires iteration over both sets.
+	 */
+	private final ReadWriteLock flushLock = new ReentrantReadWriteLock();
 
 	@Inject
-	public ChunkCache(ChunkLoader chunkLoader, ChunkSaver chunkSaver, CryptoFileSystemStats stats, BufferPool bufferPool) {
+	public ChunkCache(ChunkLoader chunkLoader, ChunkSaver chunkSaver, CryptoFileSystemStats stats, BufferPool bufferPool, ExceptionsDuringWrite exceptionsDuringWrite) {
 		this.chunkLoader = chunkLoader;
 		this.chunkSaver = chunkSaver;
 		this.stats = stats;
 		this.bufferPool = bufferPool;
-		this.chunks = CacheBuilder.newBuilder() //
+		this.exceptionsDuringWrite = exceptionsDuringWrite;
+		this.staleChunks = Caffeine.newBuilder() //
 				.maximumSize(MAX_CACHED_CLEARTEXT_CHUNKS) //
-				.removalListener(this::removeChunk) //
-				.build();
+				.evictionListener(this::evictStaleChunk) //
+				.build() //
+				.asMap();
+		this.activeChunks = new ConcurrentHashMap<>();
 	}
 
-	private Chunk loadChunk(long chunkIndex) throws IOException {
-		stats.addChunkCacheMiss();
+	/**
+	 * Overwrites data at the given index and increments the access counter of the returned chunk
+	 *
+	 * @param chunkIndex Which chunk to overwrite
+	 * @param chunkData The cleartext data
+	 * @return The chunk
+	 * @throws IllegalArgumentException If {@code chunkData}'s remaining bytes is not equal to the number of bytes fitting into a chunk
+	 */
+	public Chunk putChunk(long chunkIndex, ByteBuffer chunkData) throws IllegalArgumentException {
```

Comment:
If we call `putChunk` for an index, which is also mapped in `staleChunks`, that index is present in _both_ maps (active and stale). In my opinion, we should disallow such states and ensure the following invariant:
> At any point in time, an index is mapped either in active or stale, but not in both.

Otherwise we get cases like
1. Index 42 is only mapped to chunk c1 (42:c1) in `staleChunks`
1. Thread t_A calls `getChunk(42)`, moving (42:c1) from stale to active
1. Thread t_B calls `putChunk(42,data)`, replacing (42:c1) with (42:c2) in active
1. Thread t_A closes  c1, which leads to:
    - calling `releaseChunk(42)`
    - decrement the access counter of _c2_ (a different chunk!)
    - moving (42:c2) from active to stale
1.  Chunk c1 is "lost", its buffer won't be recylced and is only removed in the next gc run
1. Thread t_B closes c2 (because it has a reference to it), which will throw a NullPointerException, because chunk is not in active anymore

Reply:
I don't think this is an issue. Step 2 and 3 are each atomic. So regardless of which thread runs first, either `c1` xor `c2` will be in `activeChunks`.

Reply:
fwiw, you could use one mapping by [pinning](https://github.com/ben-manes/caffeine/wiki/Faq#pinning-entries) so that active chunks have a weight of zero. I don’t know if it simplifies your logic, though.

Reply:
That would indeed simplify things! I guess it should be a separate PR though 😅

Reply:
> either c1 xor c2 will be in activeChunks.

True, my mistake. Nevertheless, we discussed this in a different channel and decided to ensure this invariant.

> could use one mapping by [pinning](https://github.com/ben-manes/caffeine/wiki/Faq#pinning-entries) so that active chunks have a weight of zero

Then we would only need one map actually and don't have to move around chunks, which simpliyfies our architecture. @ben-manes Thanks for pointing that out!